### PR TITLE
Speedup cobra

### PR DESCRIPTION
**Pull request by xjunior**
## Before
```bash
bundle exec cobra ls --format=list .  766.76s user 41.04s system 99% cpu 13:34.10 total
```

### After
```bash
bundle exec cobra ls --format=list .  2.07s user 1.34s system 90% cpu 3.775 total
```

### CODE REVIEW

File: spec/fixtures/app/package.json [hunk 1]
```diff
@@ -6,7 +6,7 @@
   "license": "UNLICENSED",
   "private": true,
   "workspaces": [
-    "components/h"
+    "components/*"
```
Comment (benlangfeld) [inline spec/fixtures/app/package.json:9]: What is this for?

File: spec/fixtures/app/Gemfile [hunk 2]
```diff
@@ -9,3 +9,7 @@ path "components" do
   gem "a"
   gem "d"
 end
+
+path "../../.." do
+  gem "cobra_commander"
```
Comment (benlangfeld) [inline spec/fixtures/app/Gemfile:14]: What is this for?

File: lib/cobra_commander/dependencies/yarn/workspace.rb [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+          component.gsub(project_tag, "")
+        end
+
+        def tag(component)
+          [project_tag, component].join
+        end
+      end
+    end
+  end
+end
```
Comment (benlangfeld) [inline lib/cobra_commander/dependencies/yarn/workspace.rb:53]: Terminating newlines :)

File: lib/cobra_commander/cli.rb [hunk 4]
```diff
@@ -25,9 +25,10 @@ def do(command)
     method_option :format, default: "tree", aliases: "-f", desc: "Format (list or tree, default: list)"
     method_option :cache, default: nil, aliases: "-c", desc: CACHE_DESCRIPTION
     def ls(app_path = nil)
-      tree = maybe_cached_tree(app_path || options.app, options.cache)
```
Comment (benlangfeld) [inline lib/cobra_commander/cli.rb:28]: Dropping support for the cache?

File: lib/cobra_commander/dependencies.rb [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+require_relative "dependencies/multi_source"
+
+module CobraCommander
+  module Dependencies
+    def self.context(root, components, yarn: false, bundler: false)
+      sources = []
+      sources << Yarn.build(root) unless bundler
+      sources << Bundler.build(root, components) unless yarn
+      MultiSource.new(sources)
+    end
```
Comment (benlangfeld) [inline lib/cobra_commander/dependencies.rb:14]: Surely you mean this?

```suggestion
    def self.context(root, components, yarn: true, bundler: true)
      sources = []
      sources << Yarn.build(root) if bundler
      sources << Bundler.build(root, components) if yarn
      MultiSource.new(sources)
    end
```

Comment (benlangfeld) [reply lib/cobra_commander/dependencies.rb:14]: Also, what about this?

```suggestion
    def self.context(root, components_directory = "components", yarn: false, bundler: false)
      sources = []
      sources << Yarn.build(root) unless bundler
      sources << Bundler.build(root, components_directory) unless yarn
      MultiSource.new(sources)
    end
```

I don't see alternative values for `components_directory` being used very often.

File: lib/cobra_commander/dependencies/yarn/package_json.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+      private
+
+        def json
+          @json ||= JSON.parse(File.read(File.join(path, "package.json")))
+        end
+      end
+    end
+  end
+end
```
Comment (benlangfeld) [inline lib/cobra_commander/dependencies/yarn/package_json.rb:38]: Trailing newline :)

**Comment by xjunior** [comment]
@benlangfeld this is in good shape now, 100% compatible with the previous implementation. I'll now update nitro to use this new version in its internal scripts and clean-up the cache handling stuff. I'll then create another PR with my suggestions for a new `cobra` CLI API.

**Comment by xjunior** [comment]
@benlangfeld sure thing, but before that I'll create a nitro branch with this version and see it getting tested/deployed to make sure it's all good.

**Comment by benlangfeld** [comment]
This appears to remove support for the cache, is that right? We should remove it from the interface and docs in that case, and make note of the backward incompatibility in the changelog.

**Comment by xjunior** [comment]
@benlangfeld this does not remove the cache options or CLI, but I'm adding a deprecation now. In my next PR I'll suggest a more complete API and remove all cache options.

File: cobra_commander.gemspec [hunk 7]
```diff
@@ -11,10 +11,12 @@ Gem::Specification.new do |spec|
   spec.authors       = [
     "Ben Langfeld",
     "Garett Arrowood",
+    "Carlos Palhares",
```
Comment (garettarrowood) [inline cobra_commander.gemspec:14]: 👍

**Pull request merged**